tests/test_moe/test_kernel.py

```diff
@@ -1,98 +1,104 @@
 
 
-def run_routing(rank, world_size, port, rs=2, hidden_size=128, data_type=torch.float32, topk=1):
-    # Here we do not need TF32, since it brings absolute error on results
-    torch.backends.cuda.matmul.allow_tf32 = False
+def run_moe_cumsum():
+    test_mask = torch.tensor(
+        [
+            [0, 1, 0, 0],
+            [1, 0, 0, 0],
+            [0, 1, 0, 0],
+            [1, 0, 0, 0],
+        ],
+        dtype=torch.int32,
+    ).to("cuda")
+    out_no_kernel = moe_cumsum(test_mask, use_kernel=False)
+    out_kernel = moe_cumsum(test_mask, use_kernel=True)
+    print(out_no_kernel.dtype, out_kernel.dtype)
+    check_equal(out_no_kernel.to(torch.int32), out_kernel)
 
-    colossalai.launch(rank=rank, world_size=world_size, host="localhost", port=port, backend="nccl")
-    local_rank = dist.get_rank()
 
-    MOE_MANAGER.setup(parallel="EP")  # MOE environment initialization
-    MOE_MANAGER.reset_loss()
-    torch.manual_seed(rs + local_rank)  # set each process has different random seed
-
-    # get randomized data
+def run_moe_dispatch_combine_fwd_bwd(data_type=torch.float32, hidden_size=128, num_experts=4):
     tokens = torch.randn(
         BATCH_SIZE, hidden_size, dtype=data_type, device=get_accelerator().get_current_device(), requires_grad=True
     )
 
-    layer = SparseMLP(
-        hidden_size=hidden_size,
-        intermediate_size=hidden_size * 2,
-        num_experts=NUM_EXPERTS,
-        router_top_k=topk,
-        router_capacity_factor_train=1.0,
+    # use kernel
+    route_result_list_kernel = (
+        torch.load(f"{MOE_TENSOR_PATH}/True_4_{data_type}.pt")
+        if MOE_TENSOR_PATH
+        else torch.load(f"True_4_{data_type}.pt")
     )
-    layer = layer.to(get_accelerator().get_current_device())
-    if data_type == torch.float16:
-        layer = layer.half()
-
-    # use matrix multiplication instead of COL_MOE_KERNEL in MOE dispatch and combine
-    layer.enable_kernel = False
-    old_out = layer(tokens)
-    ech = old_out.shape
-    grad = torch.randn(ech, device=get_accelerator().get_current_device())
-    old_out.backward(grad)  # get gradient
-
-    # save all results
-    o_tk_grad = tokens.grad.data.clone()
-    o_gt_grad = layer.gate_weight.grad.data.clone()
-
-    # reset all gradients
-    tokens.grad.zero_()
-    layer.gate_weight.grad.zero_()
-
-    layer.enable_kernel = True
-    new_out = layer(tokens)  # get outputs through colossal kernel
-
+    # dispatch
+    dispatch_data_kernel = MoeDispatch.apply(tokens, *route_result_list_kernel[1:])
+    dispatch_data_kernel = dispatch_data_kernel.reshape(num_experts, -1, hidden_size)
+    # combine
+    expert_output = dispatch_data_kernel.reshape(-1, hidden_size)
+    ans_kernel = MoeCombine.apply(expert_output, *route_result_list_kernel)
+
+    # no kernel
+    route_result_list_no_kernel = (
+        torch.load(f"{MOE_TENSOR_PATH}/False_2_{data_type}.pt")
+        if MOE_TENSOR_PATH
+        else torch.load(f"False_2_{data_type}.pt")
```

Comment:
In which case is  `MOE_TENSOR_PATH` None?

Reply:
For user who test it locally, but maybe not need for other developers.